nighthawk/source/common/sequencer_impl.cc

```diff
@@ -0,0 +1,145 @@
+      ENVOY_LOG(
+          trace,
+          "SequencerImpl done processing {} operations in {} ms. (completion rate {} per second)",
+          targets_completed_,
+          std::chrono::duration_cast<std::chrono::milliseconds>(now - start_).count(), rate);
+      return;
+    } else {
+      // After the benchmark duration has exceeded, we wait for a grace period for outstanding work
+      // to wrap up. If that takes too long we warn about it and quit.
+      if (runtime - duration_ > grace_timeout_) {
+        stop();
+        ENVOY_LOG(warn,
+                  "SequencerImpl timeout waiting for due responses. Initiated: {} / Completed: {}. "
+                  "(completion ~ rate {} per second.)",
+                  targets_initiated_, targets_completed_, rate);
+        return;
+      }
+      if (from_periodic_timer) {
+        scheduleRun();
+      }
+    }
+    return;
+  }
+
+  while (rate_limiter_.tryAcquireOne()) {
+    // The rate limiter says it's OK to proceed and call the target. Let's see if the target is OK
+    // with that as well.
+    const bool ok = target_([this, now]() {
+      auto dur = time_source_.monotonicTime() - now;
+      latencyStatistic_.addValue(dur.count());
+      targets_completed_++;
+      // Immediately schedule us to check again, as chances are we can get on with the next task.
+      adhoc_timer_->enableTimer(0ms);
+    });
+    if (ok) {
+      if (blocked_) {
+        blocked_ = false;
+        blockedStatistic_.addValue((now - blocked_start_).count());
+      }
+      targets_initiated_++;
+    } else {
+      if (!blocked_) {
+        blocked_start_ = now;
+        blocked_ = true;
+      }
+      // This should only happen when we are running in closed-loop mode, which is always at the
+      // time of writing this.
+      // TODO(oschaaf): Create a specific statistic for tracking time spend here and report.
+      // Measurements will be skewed.
+      // The target wasn't able to proceed. Update the rate limiter, we'll try again later.
+      rate_limiter_.releaseOne();
+      break;
+    }
+  }
+
+  if (!from_periodic_timer) {
+    if (targets_initiated_ == targets_completed_) {
+      // We saturated the rate limiter, and there's no outstanding work.
+      // That means it looks like we are idle. Spin this event to improve
+      // accuracy. As a side-effect, this may help prevent CPU frequency scaling
+      // due to c-state changes. But on the other hand it may cause thermal throttling.
+      // TODO(oschaaf): Ideally we would have much finer grained timers instead.
+      // TODO(oschaaf): Optionize performing this spin loop.
+      platform_util_.yieldCurrentThread();
+      adhoc_timer_->enableTimer(0ms);
+    }
+  } else {
+    // Re-schedule the periodic timer if it was responsible for waking up this code.
+    scheduleRun();
+  }
+}
+
+void SequencerImpl::waitForCompletion() {
+  ASSERT(running_);
+  dispatcher_.run(Envoy::Event::Dispatcher::RunType::Block);
+  if (running_) {
```

Comment:
Good question, the only situation where `running_` equals `true` when we get here, is actually in one of the tests using mocks...
I need to think about this can actually really happen under normal conditions, maybe in an edge case, but if not it should probably be better to have `ASSERT(!running)` here.

Reply:
Yeah, anything to simplify reasoning has my +1.